bindings/wallet-core/src/c.rs

```diff
@@ -385,6 +392,125 @@ pub fn wallet_set_state(wallet: WalletPtr, value: u64, counter: u32) -> Result {
+/// the pointers are null. Mind not to put random values in or you may see
+/// unexpected behaviors.
+pub unsafe fn wallet_vote_proposal(
+    vote_plan_id: *const u8,
+    payload_type: u8,
+    index: u8,
+    num_choices: u8,
+    proposal_out: *mut ProposalPtr,
+) -> Result {
+    if vote_plan_id.is_null() {
+        return Error::invalid_input("vote_plan_id").with(NulPtr).into();
+    }
+
+    if proposal_out.is_null() {
+        return Error::invalid_input("proposal_out").with(NulPtr).into();
+    }
+
+    let payload_type = match payload_type.try_into() {
+        Ok(payload_type) => payload_type,
+        Err(err) => return Error::invalid_input("payload_type").with(err).into(),
+    };
+
+    let options = match VoteOptions::new_length(num_choices) {
+        Ok(options) => options,
+        Err(err) => return Error::invalid_input("num_choices").with(err).into(),
+    };
+
+    let vote_plan_id = std::slice::from_raw_parts(vote_plan_id, crate::vote::VOTE_PLAN_ID_LENGTH);
+    let vote_plan_id = match VotePlanId::try_from(vote_plan_id) {
+        Ok(id) => id,
+        Err(err) => return Error::invalid_input("vote_plan_id").with(err).into(),
+    };
+
+    *proposal_out = Box::into_raw(Box::new(Proposal::new(
+        vote_plan_id,
+        payload_type,
+        index,
+        options,
+    )));
+
+    Result::success()
+}
+
+/// build the vote cast transaction
+///
+/// # Errors
+///
+/// This function may fail upon receiving a null pointer or a `choice` value
+/// that does not fall within the range specified in `proposal`.
+///
+/// # Safety
+///
+/// This function dereference raw pointers. Even though the function checks if
+/// the pointers are null. Mind not to put random values in or you may see
+/// unexpected behaviors.
+pub unsafe fn wallet_vote_cast(
+    wallet: WalletPtr,
+    settings: SettingsPtr,
+    proposal: ProposalPtr,
+    choice: u8,
+    transaction_out: *mut *const u8,
+    len_out: *mut usize,
+) -> Result {
+    let wallet = if let Some(wallet) = wallet.as_mut() {
+        wallet
+    } else {
+        return Error::invalid_input("wallet").with(NulPtr).into();
+    };
+
+    let settings = if let Some(settings) = settings.as_ref() {
+        settings.clone()
+    } else {
+        return Error::invalid_input("settings").with(NulPtr).into();
+    };
+
+    let proposal = if let Some(proposal) = proposal.as_ref() {
+        proposal
+    } else {
+        return Error::invalid_input("proposal").with(NulPtr).into();
+    };
+
+    if transaction_out.is_null() {
+        return Error::invalid_input("transaction_out").with(NulPtr).into();
+    }
+    if len_out.is_null() {
+        return Error::invalid_input("len_out").with(NulPtr).into();
+    }
+
+    let choice = Choice::new(choice);
+
+    let mut transaction = match wallet.vote(settings, proposal, choice) {
+        Ok(transaction) => transaction,
+        Err(err) => return err.into(),
+    };
+
+    transaction.shrink_to_fit();
```

Comment:
Just nitpicking, I think `into_boxed_slice` could work here too.

Reply:
Should use slightly less memory too